File: frontend/src/lib/notices.js
/**
 * Server notices: what was done on other terms than the ones asked for.
 *
 * A notice is not an error. The request succeeded, and something about it was
 * decided for the caller: a validity shortened by a policy or by the issuing
 * CA's own expiry. The server puts these in `meta.notices`; without showing
 * them, the only trace is a date nobody reads until it matters.
 */

/** The notices carried by a response, or an empty array. */
export function noticesOf(response) {
  const list = response?.meta?.notices
  return Array.isArray(list) ? list.filter(Boolean) : []
}

/**
 * Show each notice as a warning.
 *
 * Returns how many were shown, so a caller can keep its success toast quiet
 * when the server had something to say instead.
 */
export function showNotices(response, showWarning) {
  const list = noticesOf(response)
  list.forEach((notice) => showWarning(notice))
  return list.length
}

File: backend/utils/notices.py
"""What the server decided differently from what was asked, said out loud.

A request that is neither refused nor honoured as written is the one case a
caller cannot see: the answer is a 200 or a 201, and the difference only shows
up later, in the expiry date of a certificate somebody is relying on. These
notices travel in the response ``meta.notices`` so the interface can show them
next to the result, and read the same way in a log.

They are not errors. A refusal goes through ``error_response``; a notice says
the thing was done, and on what terms.
"""
from __future__ import annotations

from typing import List, Optional


def validity_shortened(requested_days: int, granted_days: int,
                       reason: str) -> str:
    """The certificate was issued for less than the caller asked for."""
    return (f'Issued for {granted_days} day(s) instead of the '
            f'{requested_days} requested: {reason}.')


def policy_validity_reason(policy_name: str, max_days: int) -> str:
    """Why a policy shortened it."""
    return f'policy "{policy_name}" caps validity at {max_days} day(s)'


def issuer_expiry_reason(not_after) -> str:
    """Why the issuing CA shortened it."""
    return (f'the issuing CA certificate expires on '
            f'{not_after.strftime("%Y-%m-%d")} and no certificate may outlive it')


def collect(*notices: Optional[str]) -> List[str]:
    """The notices that are actually there, in order, without blanks."""
    return [n for n in notices if n]


def meta_with_notices(notices, meta: Optional[dict] = None) -> Optional[dict]:
    """``meta`` carrying *notices*, or the meta unchanged when there are none."""
    kept = [n for n in (notices or []) if n]
    if not kept:
        return meta
    out = dict(meta or {})
    out['notices'] = kept
    return out


__all__ = [
    'validity_shortened',
    'policy_validity_reason',
    'issuer_expiry_reason',
    'collect',
    'meta_with_notices',
]
